ql/src/java/org/apache/hadoop/hive/ql/exec/MoveTask.java

```diff
@@ -97,6 +101,40 @@ public MoveTask() {
     super();
   }
 
+  public void flattenUnionSubdirectories(Path sourcePath) throws HiveException {
+    try {
+      FileSystem fs = sourcePath.getFileSystem(conf);
+      LOG.info("Checking " + sourcePath + " for subdirectories to flatten");
+      Set<Path> unionSubdirs = new HashSet<>();
+      if (fs.exists(sourcePath)) {
```

Comment:
This is a costly call. We have a same result with catching the relevant exception in `listFiles`, and we can save one FileSystem.exists call.